### PR TITLE
feat: add a ImmutableBuffer

### PR DESCRIPTION
**Pull request by kristian**
ImmutableBuffer throws a UnsupportedOperationException for anything that writes to the buffer. This disadvantage is compensated by the advantage the ImmutableBuffer can be transfered over the local event bus very efficiently without beeing copied.

### CODE REVIEW

File: src/test/java/io/neonbee/test/base/NeonBeeTestBase.java [hunk 1]
```diff
@@ -113,26 +143,42 @@ public void setUp(Vertx vertx, VertxTestContext testContext, TestInfo testInfo)
...
+                    DeploymentOptions serverVerticleOptions =
+                            readDeploymentOptions(ServerVerticle.class, workingDirPath);
 
-                Optional.ofNullable(provideUserPrincipal(testInfo)).map(userPrincipal -> {
-                    // Replace current ServerVerticle with a dummy ServerVerticle that also has a dummy AuthHandler to
-                    // provide the user principal specified in the provideUserPrincipal method
-                    ServerVerticle dummyServerVerticle = createDummyServerVerticle(testInfo);
+                    try {
+                        // just to be sure, try to fetch a new port to use
+                        options.setServerPort(SystemHelper.getFreePort());
```
Comment (pk-work) [inline src/test/java/io/neonbee/test/base/NeonBeeTestBase.java:151]: The downside of this is, that the ServerVerticleConfig  in /config folder is not longer in sync with the instantiated object .. Maybe we should write the "new" config also back into the /config foler.

**Review by pk-work — CHANGES_REQUESTED** [review]

File: src/test/java/io/neonbee/NeonBeeExtension.java [hunk 2]
```diff
@@ -107,25 +108,56 @@ private Object unpack(Object object) {
...
+            @NeonBeeInstanceConfiguration Void nothing) {
+        try {
+            return NeonBeeExtension.class.getDeclaredMethod("internalDefaultInstanceConfiguration", Void.class)
+                    .getParameters()[0].getAnnotation(NeonBeeInstanceConfiguration.class);
+        } catch (NoSuchMethodException | SecurityException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static NeonBeeOptions.Mutable defaultOptions() {
```
[the remark this answers is not in the document]
Comment (kristian) [reply src/test/java/io/neonbee/NeonBeeExtension.java:127]: I moved it into an own "OptionsHelper" class

File: src/test/java/io/neonbee/job/JobVerticleTest.java [hunk 3]
```diff
@@ -93,6 +98,12 @@ public String deploymentID() {
         }
     }
 
+    @Override
+    protected void adaptOptions(TestInfo testInfo, NeonBeeOptions.Mutable options) {
+        options.setActiveProfiles(List.of(ALL, NO_WEB));
+        options.setDisableJobScheduling(false);
```
[the remark this answers is not in the document]
Comment (kristian) [reply src/test/java/io/neonbee/job/JobVerticleTest.java:102]: Because WatchVerticle is no JobVerticle :)

File: src/test/java/io/neonbee/test/helper/SystemHelper.java [hunk 4]
```diff
@@ -2,19 +2,48 @@
...
 public final class SystemHelper {
+    // the ephemeral port range, as suggested by the IANA (https://www.iana.org/assignments/port-numbers)
+    private static final Range<Integer> PORT_RANGE = Range.closed(49152, 65535);
+
+    private static final Iterator<Integer> PORTS =
+            ContiguousSet.create(PORT_RANGE, DiscreteDomain.integers()).iterator();
 
     /**
-     * This method tries to find a free port on the system and returns it.
+     * This method tries to find a free ephemeral port on the system and returns it. This method will never return the
```
[the remark this answers is not in the document]
Comment (kristian) [reply src/test/java/io/neonbee/test/helper/SystemHelper.java:21]: But it does pick a random (?) port, the new implementation will guarantee, that for one test run (in one JVM) no same port is returned twice.

File: src/test/java/io/neonbee/test/listeners/StaleVertxChecker.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            // perfect! we don't need to bother ourself!
+            LOGGER.info("No stale Vert.x instances found");
+            return;
+        } else if (vertxInstances.contains(null)) {
+            LOGGER.warn("Could not determine Vert.x instance for all Vert.x threads");
+        }
+
+        // try to find instances associated to this test or any instances that we cannot associate to any test
+        Set<Vertx> unassociatedVertxInstances = new HashSet<>(), associatedVertxInstances = new HashSet<>(), // NOPMD
+                notAssociatedVertxInstances = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (kristian) [reply src/test/java/io/neonbee/test/listeners/StaleVertxChecker.java:123]: No, this does not make sense to me. Then we should adapt the PMD ruling.

If I want to define multiple sets with the same generic type, which MUST be of the same generic type, it doesn't make sense to declare it multiple times, which is the bigger code smell to me, because it assumes that the sets can be of different generic type, which is not true:

Set<Vertx> unassociatedVertxInstances = new HashSet<>();

Set<Vertx> associatedVertxInstances = new HashSet<>();

Set<Vertx> notAssociatedVertxInstances = new HashSet<>();

Thus it is in one declaration, because it makes it clear: "These sets MUST be compatible to each other".

Please let me know if the PMD rule should be adapted, or if you are fine with the NOPMD here.

File: build.gradle [hunk 6]
```diff
@@ -9,16 +9,16 @@ plugins {
...
-    id 'com.github.ksoichiro.console.reporter' version '0.6.2'
-    id 'se.bjurr.violations.violations-gradle-plugin' version '1.48'
-    id 'com.github.johnrengelman.shadow' version '6.1.0'
-    id 'org.sonarqube' version '3.1.1'
-    id 'team.yi.semantic-gitlog' version '0.5.13'
+    id 'com.github.ksoichiro.console.reporter' version '0.6.3'
+    id 'se.bjurr.violations.violations-gradle-plugin' version '1.50.16'
+    id 'com.github.johnrengelman.shadow' version '7.0.0'
+    id 'org.sonarqube' version '3.3'
+    //id 'team.yi.semantic-gitlog' version '0.5.17' // isn't compatible with Gradle 7.2 yet
```
[the remark this answers is not in the document]
Comment (kristian) [reply build.gradle:21]: At some point in time I updated everything, because I did assume it is a bug in the parallel execution of JUnit which was quite new. As JUnit parallel execution is build into Gradle I had to update gradle.

The plugin is only used for generating the changelog as far as I can tell. Executing the plugin with a lower Gradle is much less effort, than doing the whole Gradle update again. Thus I would argue we update it, as soon as it is available for Gradle 7. If the plugin will not be updated, we anyways have to check for a different plugin, as Gradle 8 is already around the corner and the Gradle 7 update would have been due anyways.

File: docs/internal/howto_debug_github_actions.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+This document describes how to hook into the GitHub actions workflow if required using a remote SSH session.
+
+Add the following workflow step into any position of the `.github/workflows` file of your choice, e.g. [`voter.yml`](../../.github/workflows/voter.yml):
+
+```
+- name: Setup tmate session
+  uses: mxschmitt/action-tmate@v3
+```
+
+It causes a [tmate](https://tmate.io/) session to be spun up (attention it is a public session, anybody could hook into it, when they monitor our build logs, protection would be possible with options of the `action-tmate`, however for an open source project like ours it is kind of irrelevant). The action *will wait* until you connect to the session via SSH and create a `continue` file (e.g. by using `touch continue`). Wait for the step to execute, checking the build logs you should see an SSH info like:
```
[the remark this answers is not in the document]
Comment (kristian) [reply docs/internal/howto_debug_github_actions.md:12]: Which credentials do you mean?

File: src/test/java/io/neonbee/test/base/NeonBeeTestBase.java (hunk 1 above)
Comment (kristian) [reply src/test/java/io/neonbee/test/base/NeonBeeTestBase.java:151]: Yes I knew. I tolerated this, because this is only covering an edge case, which should normally not ocour. However all methods which read the port, anyways read it from the options. The config is just there "in addition". So we should be good with this.

**Comment by kristian** [comment]
Thanks for the review @pk-work. I have three open discussions, would you mind providing your feedback? Thanks.

File: src/test/java/io/neonbee/job/JobVerticleTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pk-work) [reply src/test/java/io/neonbee/job/JobVerticleTest.java:102]: LoL good point! Maybe it should become one in the future.

File: build.gradle (hunk 6 above)
[the remark this answers is not in the document]
Comment (pk-work) [reply build.gradle:21]: I agree, haven't thought about running the changelog task with an older version. This makes sense.but how can we achieve this?

File: docs/internal/howto_debug_github_actions.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pk-work) [reply docs/internal/howto_debug_github_actions.md:12]: I mean credentials like this: https://github.com/SAP/neonbee/blob/8fce18285048a1c560af60aa053f766353ec775e/.github/workflows/voter.yml#L55

File: build.gradle (hunk 6 above)
[the remark this answers is not in the document]
Comment (kristian) [reply build.gradle:21]: simply change the wrapper version and run the task, just work just fine :)

File: docs/internal/howto_debug_github_actions.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kristian) [reply docs/internal/howto_debug_github_actions.md:12]: Mhm, I don't know... let's be sure and always remove tmate again if debugging is done

I will provide a comment in the docs

**Review by s4heid — CHANGES_REQUESTED** [review]
Congrats, @kristian for this valuable change 🚀  there are only a few small things I would like to adjust.

File: src/main/java/io/neonbee/internal/verticle/WatchVerticle.java [hunk 8]
```diff
@@ -150,12 +152,19 @@ public WatchVerticle(Path watchPath, long interval, TimeUnit unit, boolean paral
...
+
         try {
             watcher = watchPath.getFileSystem().newWatchService();
         } catch (IOException e) {
             startPromise.fail(e);
             return;
         }
+
         (handleExisting ? handleExistingFiles(watchPath) : registerWatchKey(watchPath)).compose(
                 v -> Future.<Counter>future(promise -> getVertx().sharedData().getLocalCounter(counterName, promise)))
```
[the remark this answers is not in the document]
Comment (kristian) [reply src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:169]: `getVertx()` of `AbstractVerticle` will internally return the reference to this.vertx, however in abstract classes, or classes that extension is intended (e.g. WatchVerticle) it is considered a code smell to use the direct reference to a field, if a getter is available. This enables for instance overriding getVertx in tests to return a mocked Vert.x instance instead.

File: src/main/java/io/neonbee/internal/verticle/WatchVerticle.java [hunk 9]
```diff
@@ -150,12 +152,19 @@ public WatchVerticle(Path watchPath, long interval, TimeUnit unit, boolean paral
 
     @Override
     public void start(Promise<Void> startPromise) {
+        Vertx vertx = getVertx();
+        if (NeonBee.get(vertx).getOptions().doNotWatchFiles()) {
+            startPromise.fail("Should not watch files");
```
[the remark this answers is not in the document]
Comment (kristian) [reply src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:157]: You are right, I changed it to a similar approach that we had in JobVerticle and undeploy the verticle immediately after start.

File: src/test/java/io/neonbee/NeonBeeInstanceConfiguration.java [hunk 10]
```diff
@@ -32,6 +32,8 @@
 
     boolean ignoreClassPath() default true;
 
+    boolean doNotWatchFiles() default true;
```
[the remark this answers is not in the document]
Comment (kristian) [reply src/test/java/io/neonbee/NeonBeeInstanceConfiguration.java:37]: this class is only used in / for tests. therefore it is also in src/test/.... In non-tests the default is actually false, as you would expect.

**Comment by kristian** [comment]
I added all the changes requested. @s4heid, @pk-work, I just added one more `feat:` as the previous to last commit. Could you please do a code review specifically for this one? The rest is changed as you reviewed it already. Afterwards we should be able to finally merge this beast! :)

File: src/main/java/io/neonbee/internal/verticle/WatchVerticle.java [hunk 11]
```diff
@@ -154,7 +156,13 @@ public WatchVerticle(Path watchPath, long interval, TimeUnit unit, boolean paral
     public void start(Promise<Void> startPromise) {
         Vertx vertx = getVertx();
         if (NeonBee.get(vertx).getOptions().doNotWatchFiles()) {
```
Comment (pk-work) [inline src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:158]: In my opionion the we shouldn't check in WatchVerticle if the NeonBee option "doNotWatchFiles" ist set or not. It is a NeonBeeOption and should only change the behavior of NeonBee. If I want to deploy a Module, which contains a WatchVerticle I don't want that this NeonBee option prevents it.

We have 2 verticles that extends the WatchVerticle DeployerVerticle and ModelRefresher. If NeonBee option "doNotWatchFiles" is set, we simply shouldn't deploy this verticles.

Comment (kristian) [reply src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:158]: I see your point, however then we have a different behaviour in `JobVerticle` than we have in `WatchVerticle`. And if you don't want to have _any_ files watched, that's what the NeonBee option is for. The question is: Should we "enforce" it, then I would say, adding it to the `WatchVerticle` is kind of how we did it in the `JobVerticle` as well, or should we _not_ enforce it, then I would agree to your option. What do you think?

Comment (pk-work) [reply src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:158]: Maybe its the name of the option. I thought we introduced the option "doNotWatchFiles" with the idea to stop NeonBee watching the folders "models" and "verticles". Or did we had another idea in mind?

If this was the idea, I wouldn't enforce it. I got your point with the similarity to "shouldDisableJobScheduling", but then maybe we simply should rename the option into "doNotWatchModules" and try to "remove" the models folder as soon as possible. Or rename it to "doNotWatchResources" or "doNotWatchModelsAndModules".

To be honest in my opinion "shouldDisableJobScheduling" is also not really required. I'd prefer disabling the "unwanted" JobVerticles via their config.

Comment (kristian) [reply src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:158]: Well both options are particularly useful in tests, because in many / most tests you don't want to have files watch nor you want job verticles, Thus the default option for tests has been set to "true / true" so no jobs are started and no files will be watched. Maybe we should move this debate to an issue instead to have it in our backlog and resolve this conversation?

Comment (pk-work) [reply src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:158]: I don't know, because this change then introduces a behavior which maybe is unwanted and needs to be removed afterwards. On the other hand this PR is so big in the meanwhile, that it is hard to review. So maybe it makes sense to merge it now as it is and do the "removal" maybe later.

I like none of these options, so you can decide :)

Comment (kristian) [reply src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:158]: I would like to merge this change *now* :D if you permit! 👯

Then this beast and the whole chapter on non-running tests is hopefully and FINALLY closed!

**Comment by kristian** [comment]
The change is now DONE. Can we merge, please? 👍

**Review by pk-work — APPROVED** [review]
LGTM To be honest I haven't had yet a final look again, but I think there were only small fixes.

**Comment by kristian** [comment]
@pk-work, @s4heid I will click the button now! any further changes will go into a n ew PR!! 👍  THIS ONE IS DONE AND OVER!! Final agreement?

**Review by s4heid — APPROVED** [review]

**Comment by kristian** [comment]
It's done! 👍 Thanks a lot to both of you! @gedack, feel free to rebase your other pull requests now, which should fix the voter for them finally! *cheer*
